marketplace-service/src/main/java/com/axonivy/market/service/impl/MetadataServiceImpl.java

```diff
@@ -0,0 +1,323 @@
+      updateMavenArtifactVersionFromMetadata(artifactVersionCache, metadata);
+      updateContentsFromNonMatchVersions(releasedVersions, metadata);
+    }
+  }
+
+  public void syncAllProductsMetadata() {
+    List<Product> products = productRepo.getAllProductsWithIdAndReleaseTagAndArtifact();
+    log.warn("**MetadataService: Start to sync version for {} product(s)", products.size());
+    for (Product product : products) {
+      // Set up cache before sync
+      String productId = product.getId();
+      Set<Metadata> metadataSet = new HashSet<>(metadataRepo.findByProductId(product.getId()));
+      MavenArtifactVersion artifactVersionCache = mavenArtifactVersionRepo.findById(product.getId()).orElse(
+          MavenArtifactVersion.builder().productId(productId).productArtifactsByVersion(
+              new HashMap<>()).additionalArtifactsByVersion(new HashMap<>()).build());
+      MetadataSync syncCache = metadataSyncRepo.findById(product.getId()).orElse(
+          MetadataSync.builder().productId(product.getId()).syncedVersions(new HashSet<>()).build());
+      Set<Artifact> artifactsFromNewTags = new HashSet<>();
+
+      // Find artifacts from unhandled tags
+      List<String> nonSyncedVersionOfTags = VersionUtils.removeSyncedVersionsFromReleasedVersions(
+          product.getReleasedVersions(), syncCache.getSyncedVersions());
+      if (ObjectUtils.isNotEmpty(nonSyncedVersionOfTags)) {
+        artifactsFromNewTags.addAll(getArtifactsFromNonSyncedVersion(product.getId(), nonSyncedVersionOfTags));
+        log.info("**MetadataService: New tags detected: {} in product {}", nonSyncedVersionOfTags.toString(),
+            productId);
+      }
+
+      // Sync versions from maven & update artifacts-version table
+      List<Artifact> additionalArtifactFromMeta = product.getArtifacts();
+      metadataSet.addAll(MavenUtils.convertArtifactsToMetadataSet(artifactsFromNewTags, productId));
+      metadataSet.addAll(
+          MavenUtils.convertArtifactsToMetadataSet(new HashSet<>(additionalArtifactFromMeta), productId));
+
+      if (CollectionUtils.isEmpty(metadataSet)) {
+        continue;
+      }
+      artifactVersionCache.setAdditionalArtifactsByVersion(new HashMap<>());
+      updateMavenArtifactVersionData(product.getReleasedVersions(), metadataSet, artifactVersionCache);
+
+      // Persist changed
+      syncCache.getSyncedVersions().addAll(nonSyncedVersionOfTags);
+      metadataSyncRepo.save(syncCache);
+      mavenArtifactVersionRepo.save(artifactVersionCache);
+      metadataRepo.saveAll(metadataSet);
+    }
+    log.warn("**MetadataService: version sync finished");
+  }
+
+  private void updateContentsFromNonMatchVersions(List<String> releasedVersions,
+      Metadata metadata) {
+    Set<String> nonMatchSnapshotVersions = new HashSet<>();
+    List<ProductModuleContent> productModuleContents = new ArrayList<>();
+    getNonMatchSnapshotVersions(releasedVersions, metadata, nonMatchSnapshotVersions);
```

Comment:
please return Set<String> from getNonMatchSnapshotVersions, then set value for nonMatchSnapshotVersions